### ISSUE TITLE
DB null layout render doesn't work in whenEmpty

### ISSUE THREAD

**Issue by leonardder** [NLog/NLog#4070]
**NLog version**: 4.7.2
**Platform**: .Net 4.8
**Current NLog config**

```xml
<?xml version="1.0" encoding="utf-8" ?>
<nlog internalLogLevel="Warn"
  internalLogFile="C:\Temp\NLog.Internal.txt"
  internalLogToConsole="true"
  autoReload="true" 
  throwExceptions="true"
  xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <targets async="true">
    <target xsi:type="Database" name="DbDefault"
            connectionStringName = "frameworkLogging"
            commandText = "INSERT INTO Logs ([Date], [Level], [Logger], [Message], [Exception], [LogGuid], [Username], [Source], [LogGroupingIdentifier]) VALUES (@log_date, @level, @logger, @message, @exception, @logGuid, @username, @source, @logGroupingIdentifier)" keepConnection="true">
      ...
      <parameter name="@exception" layout="${exception:format=tostring:whenEmpty=${db-null}}" />
      ...
    </target>
    ...
</nlog>
```

- What is the current result?
When exception is empty/null, an empty string.
- What is the expected result?
When exception is empty/null, a NULL value is written to the database.
- Did you checked the [Internal log](https://github.com/NLog/NLog/wiki/Internal-Logging)?
trace: `2020-08-04 10:27:20.6864 Trace   DatabaseTarget: Parameter: '@exception' = '' (String)`
- Are there any workarounds? no
- Is there a version in which it did work?
Not that I know of
- Can you help us by writing an unit test?
Not sure

**Comment by snakefoot** [comment]
What happens if you add `parameterType="System.Object"` :

```xml
<parameter name="@exception" layout="${exception:format=tostring:whenEmpty=${db-null}}" parameterType="System.Object" />
```

**Comment by leonardder** [comment]
Interesting, that works like a charm.
I assume that could also fix cases where I try to log a `system.Guid?` to a nullable guid column.
Would it be possible within the scope of this issue to document the use of system.Object for nullable columns? Or may be it's clearer to add an additional argument to to parameter, like:

`<parameter name="@exception" layout="${exception:format=tostring:whenEmpty=${db-null}}" allowNullValues="true" />`

**Comment by snakefoot** [comment]
There are two DatabaseTarget parameter properties:

- `dbType="DbType"` - What type to send to the database. 
- `parameterType="Type"` - What type to convert into before sending to the database (Default String, but if having specified `dbType=` then it tries to guess a better value).

When `parameterType` converts to string, then it also translates `null` into `string.Empty`. When using `parameterType="System.Object"` then it means do not perform any conversion (Take the raw output).

NLog 4.6 changed DatabaseTarget from only supporting string to support custom DbType and ParameterType. But it would be breaking change if starting to take the non-string-road by default (Not having configured `dbType` or `parameterType`).

**Comment by snakefoot** [comment]
Have no opinion about adding a `allowDbNull`-property so when `null` or `string.empty` then it automatically coalesce into DbNull. Removing the need to specify `whenEmpty=${db-null}`.

**Comment by snakefoot** [comment]
Resolved by #4076
